src/unit.c

```diff
@@ -1922,3 +1543,400 @@ SEXP validUnits(SEXP units)
     }
     return answer;
 }
+SEXP validData(SEXP data, SEXP validUnits, int n) {
+	int nData = LENGTH(data);
+	int nUnit = LENGTH(validUnits);
+	int *pValidUnits = INTEGER(validUnits);
+	int dataCopied = 0;
+	
+	if (nData != 1 && nData != n) {
+		error(_("data must be either NULL, have length 1, or match the length of the final unit vector"));
+	}
+	
+	for (int i = 0; i < nUnit; i++) {
+		SEXP singleData = VECTOR_ELT(data, i % nData);
+		int singleUnit = pValidUnits[i % nUnit];
+		int unitIsString = isStringUnit(singleUnit);
+		int unitIsGrob = isGrobUnit(singleUnit);
+		
+		if (unitIsString && !Rf_isString(singleData) && !Rf_isLanguage(singleData)) {
+			error(_("no string supplied for 'strwidth/height' unit"));
+		}
+		if (unitIsGrob) {
+			if (!Rf_inherits(singleData, "grob") && !Rf_inherits(singleData, "gPath") && !Rf_isString(singleData)) {
+				error(_("no 'grob' supplied for 'grobwidth/height' unit"));
+			}
+			if (Rf_isString(singleData)) {
+				if (!dataCopied) {
+					data = PROTECT(shallow_duplicate(data));
+					dataCopied = 1;
+				}
+				SEXP fcall = PROTECT(lang2(install("gPath"), singleData));
+				singleData = eval(fcall, R_gridEvalEnv);
+				SET_VECTOR_ELT(data, i % nData, singleData);
+				UNPROTECT(1);
+			}
+			if (Rf_inherits(singleData, "gPath")) {
+				SEXP fcall = PROTECT(lang2(install("depth"), singleData));
+				SEXP depth = PROTECT(eval(fcall, R_gridEvalEnv));
+				int tooDeep = INTEGER(depth)[0] > 1;
+				UNPROTECT(2);
+				if (tooDeep) {
+					error(_("'gPath' must have depth 1 in 'grobwidth/height' units"));
+				}
+			}
+		}
+		if (!unitIsString && !unitIsGrob && singleData != R_NilValue) {
+			error(_("non-NULL value supplied for plain unit"));
+		}
+	}
+	UNPROTECT(dataCopied);
+	return data;
+}
+void makeSimpleUnit(SEXP values, SEXP unit) {
+	setAttrib(values, install("unit"), unit);
+	SEXP classes = PROTECT(allocVector(STRSXP, 2));
+	SET_STRING_ELT(classes, 0, mkChar("simpleUnit"));
+	SET_STRING_ELT(classes, 1, mkChar("unit"));
+	classgets(values, classes);
+	UNPROTECT(1);
+}
+SEXP constructUnits(SEXP amount, SEXP data, SEXP unit) {
+	int nAmount = LENGTH(amount);
+	int nData = LENGTH(data);
+	int nUnit = LENGTH(unit);
+	SEXP valUnits = PROTECT(validUnits(unit));
+	if (nUnit == 1) {
```

Comment:
This condition appears to generate an old-style unit that is just a numeric vector with "unit" attribute.  Isn't that what makeSimpleUnit() does ?  I thought the idea was to eliminate those ?

Reply:
BTW, hope it is ok to make these comments piecemeal like this.  It is the best way for me to avoid forgetting to make a comment ! :)

Reply:
See my general comment about the `simpleUnit` class

Reply:
Happy with this now.  Thanks